File: scripts/generate-baseline-data.py
#!/usr/bin/env python3

"""
Generate realistic AI baseline laser cutting parameters for Supabase seeding

Generates 500-600 parameter sets across 20 materials and thicknesses
with scientifically reasonable defaults based on material properties.

Usage:
    python3 generate-baseline-data.py              # Generate both JSON + SQL
    python3 generate-baseline-data.py --dry-run    # Preview only
    python3 generate-baseline-data.py --json       # JSON only
    python3 generate-baseline-data.py --sql        # SQL only
    python3 generate-baseline-data.py --output /path/to/dir
"""

import json
import random
import sys
from datetime import datetime, timedelta
from pathlib import Path
from typing import Dict, List, Tuple

# Material properties & cutting parameters
# Based on industry standards for fiber lasers
MATERIAL_CONFIGS = {
    # --- EXISTING MATERIALS (unchanged) ---
    'Stainless Steel 304': {
        'category': 'stainless',
        'thicknesses': [0.5, 1, 2, 3, 5, 8],
        'gas_types': ['O2', 'N2', 'air'],
        'power_ranges': {
            0.5: (25, 80),
            1: (40, 90),
            2: (40, 90),
            3: (60, 100),
            5: (60, 100),
            8: (60, 100),
        },
        'speed_ranges': {
            0.5: (800, 1400),
            1: (400, 800),
            2: (400, 800),
            3: (150, 400),
            5: (150, 400),
            8: (150, 400),
        },
    },
    'Mild Steel': {
        'category': 'mild_steel',
        'thicknesses': [0.5, 1, 2, 3, 5, 8],
        'gas_types': ['O2', 'air'],
        'power_ranges': {
            0.5: (20, 75),
            1: (35, 85),
            2: (35, 85),
            3: (50, 95),
            5: (50, 95),
            8: (50, 95),
        },
        'speed_ranges': {
            0.5: (900, 1600),
            1: (500, 1000),
            2: (500, 1000),
            3: (200, 500),
            5: (200, 500),
            8: (200, 500),
        },
    },
    'Aluminum': {
        'category': 'aluminum',
        'thicknesses': [0.5, 1, 2, 3, 5],
        'gas_types': ['N2', 'air'],
        'power_ranges': {
            0.5: (15, 70),
            1: (30, 80),
            2: (30, 80),
            3: (45, 90),
            5: (45, 90),
        },
        'speed_ranges': {
            0.5: (1000, 1800),
            1: (600, 1200),
            2: (600, 1200),
            3: (300, 700),
            5: (300, 700),
        },
    },
    'Acrylic': {
        'category': 'non_metal',
        'thicknesses': [1, 3, 5, 8, 10],
        'gas_types': ['air'],
        'power_ranges': {
            1: (10, 40),
            3: (10, 40),
            5: (15, 60),
            8: (15, 60),
            10: (15, 60),
        },
        'speed_ranges': {
            1: (200, 600),
            3: (200, 600),
            5: (100, 300),
            8: (100, 300),
            10: (100, 300),
        },
    },
    'Copper': {
        'category': 'copper',
        'thicknesses': [0.5, 1, 2, 3],
        'gas_types': ['N2'],
        'power_ranges': {
            0.5: (30, 85),
            1: (50, 95),
            2: (50, 95),
            3: (70, 100),
        },
        'speed_ranges': {
            0.5: (300, 800),
            1: (150, 400),
            2: (150, 400),
            3: (80, 250),
        },
    },
    'Leather': {
        'category': 'non_metal',
        'thicknesses': [1, 2, 3, 5],
        'gas_types': ['air'],
        'power_ranges': {
            1: (5, 25),
            2: (5, 25),
            3: (10, 40),
            5: (10, 40),
        },
        'speed_ranges': {
            1: (100, 400),
            2: (100, 400),
            3: (50, 200),
            5: (50, 200),
        },
    },
    # --- NEW MATERIALS ---
    'Stainless Steel 316': {
        'category': 'stainless',
        'thicknesses': [0.5, 1, 2, 3, 5, 8],
        'gas_types': ['N2', 'O2'],
        'power_ranges': {
            0.5: (30, 80),
            1: (45, 92),
            2: (50, 95),
            3: (65, 100),
            5: (70, 100),
            8: (80, 100),
        },
        'speed_ranges': {
            0.5: (750, 1350),
            1: (380, 780),
            2: (250, 600),
            3: (140, 380),
            5: (80, 250),
            8: (40, 150),
        },
    },
    'Stainless Steel 430': {
        'category': 'stainless',
        'thicknesses': [0.5, 1, 2, 3, 5],
        'gas_types': ['N2', 'O2'],
        'power_ranges': {
            0.5: (25, 75),
            1: (40, 88),
            2: (45, 92),
            3: (60, 98),
            5: (65, 100),
        },
        'speed_ranges': {
            0.5: (850, 1500),
            1: (420, 850),
            2: (280, 650),
            3: (160, 420),
            5: (90, 280),
        },
    },
    'Carbon Steel': {
        'category': 'mild_steel',
        'thicknesses': [0.5, 1, 2, 3, 5, 8, 10, 12],
        'gas_types': ['O2', 'air'],
        'power_ranges': {
            0.5: (20, 70),
            1: (30, 80),
            2: (40, 88),
            3: (50, 92),
            5: (55, 95),
            8: (65, 100),
            10: (75, 100),
            12: (85, 100),
        },
        'speed_ranges': {
            0.5: (1000, 1800),
            1: (600, 1200),
            2: (400, 900),
            3: (250, 600),
            5: (150, 400),
            8: (80, 250),
            10: (50, 180),
            12: (30, 120),
        },
    },
    'Galvanized Steel': {
        'category': 'mild_steel',
        'thicknesses': [0.5, 1, 2, 3, 5],
        'gas_types': ['O2', 'N2', 'air'],
        'power_ranges': {
            0.5: (25, 75),
            1: (35, 85),
            2: (45, 90),
            3: (55, 95),
            5: (60, 100),
        },
        'speed_ranges': {
            0.5: (850, 1500),
            1: (450, 950),
            2: (300, 700),
            3: (180, 450),
            5: (100, 300),
        },
    },
    'Brass': {
        'category': 'copper',
        'thicknesses': [0.5, 1, 2, 3],
        'gas_types': ['N2', 'air'],
        'power_ranges': {
            0.5: (30, 80),
            1: (45, 90),
            2: (55, 95),
            3: (65, 100),
        },
        'speed_ranges': {
            0.5: (400, 900),
            1: (200, 500),
            2: (120, 350),
            3: (70, 200),
        },
    },
    'Titanium': {
        'category': 'exotic',
        'thicknesses': [0.5, 1, 2, 3],
        'gas_types': ['N2', 'Ar'],
        'power_ranges': {
            0.5: (25, 75),
            1: (40, 88),
            2: (55, 95),
            3: (65, 100),
        },
        'speed_ranges': {
            0.5: (600, 1200),
            1: (350, 750),
            2: (180, 450),
            3: (100, 300),
        },
    },
    'Inconel': {
        'category': 'exotic',
        'thicknesses': [1, 2, 3],
        'gas_types': ['N2', 'Ar'],
        'power_ranges': {
            1: (55, 95),
            2: (65, 100),
            3: (75, 100),
        },
        'speed_ranges': {
            1: (200, 500),
            2: (100, 300),
            3: (50, 180),
        },
    },
    'Hastelloy': {
        'category': 'exotic',
        'thicknesses': [1, 2, 3],
        'gas_types': ['N2', 'Ar'],
        'power_ranges': {
            1: (55, 95),
            2: (70, 100),
            3: (80, 100),
        },
        'speed_ranges': {
            1: (180, 450),
            2: (90, 280),
            3: (45, 160),
        },
    },
    'Plywood': {
        'category': 'non_metal',
        'thicknesses': [3, 5, 8, 12, 18],
        'gas_types': ['air'],
        'power_ranges': {
            3: (15, 45),
            5: (20, 55),
            8: (30, 65),
            12: (40, 75),
            18: (55, 90),
        },
        'speed_ranges': {
            3: (300, 700),
            5: (200, 500),
            8: (120, 350),
            12: (70, 200),
            18: (30, 120),
        },
    },
    'MDF': {
        'category': 'non_metal',
        'thicknesses': [3, 5, 8, 12],
        'gas_types': ['air'],
        'power_ranges': {
            3: (12, 40),
            5: (18, 50),
            8: (25, 60),
            12: (35, 75),
        },
        'speed_ranges': {
            3: (350, 800),
            5: (250, 600),
            8: (150, 400),
            12: (80, 250),
        },
    },
    'Polycarbonate': {
        'category': 'non_metal',
        'thicknesses': [1, 2, 3, 5],
        'gas_types': ['air', 'N2'],
        'power_ranges': {
            1: (8, 30),
            2: (12, 40),
            3: (18, 50),
            5: (25, 65),
        },
        'speed_ranges': {
            1: (300, 700),
            2: (200, 500),
            3: (120, 350),
            5: (70, 220),
        },
    },
    'Rubber': {
        'category': 'non_metal',
        'thicknesses': [1, 2, 3, 5],
        'gas_types': ['air'],
        'power_ranges': {
            1: (8, 30),
            2: (12, 38),
            3: (18, 48),
            5: (25, 60),
        },
        'speed_ranges': {
            1: (150, 500),
            2: (100, 350),
            3: (60, 250),
            5: (30, 150),
        },
    },
    'Felt': {
        'category': 'non_metal',
        'thicknesses': [1, 2, 3],
        'gas_types': ['air'],
        'power_ranges': {
            1: (5, 20),
            2: (8, 28),
            3: (12, 35),
        },
        'speed_ranges': {
            1: (200, 600),
            2: (150, 450),
            3: (100, 350),
        },
    },
    'Cork': {
        'category': 'non_metal',
        'thicknesses': [2, 3, 5],
        'gas_types': ['air'],
        'power_ranges': {
            2: (8, 25),
            3: (12, 35),
            5: (18, 45),
        },
        'speed_ranges': {
            2: (200, 550),
            3: (150, 400),
            5: (80, 250),
        },
    },
}


def get_random_int(min_val: int, max_val: int) -> int:
    """Generate random integer between min and max."""
    return random.randint(min_val, max_val)


def get_random_float(min_val: float, max_val: float, decimals: int = 1) -> float:
    """Generate random float with specified decimal places."""
    val = random.uniform(min_val, max_val)
    return round(val, decimals)


def random_element(arr: List) -> any:
    """Return random element from list."""
    return random.choice(arr)


def generate_date_in_last_n_days(days: int) -> str:
    """Generate ISO datetime string for date in last N days."""
    now = datetime.utcnow()
    days_ago = random.uniform(0, days)
    date = now - timedelta(days=days_ago)
    return date.isoformat() + 'Z'


def generate_parameter_set(material: str, thickness: float, config: Dict) -> Dict:
    """Generate a single parameter set for material/thickness combo."""

    power_min, power_max = config['power_ranges'][thickness]
    speed_min, speed_max = config['speed_ranges'][thickness]
    gas_type = random_element(config['gas_types'])

    # Focus position: typically -5 to +5mm
    focus_position = get_random_float(-5, 5, 1)

    # Line interval: typically 0.05-0.15mm
    line_interval = get_random_float(0.05, 0.15, 3)

    # Nozzle diameter: typically 1-4mm
    nozzle_diameter = get_random_float(1, 4, 1)

    # Nozzle distance: typically 0.5-2mm
    nozzle_distance = get_random_float(0.5, 2, 1)

    # Gas pressure varies by type
    if gas_type == 'O2':
        gas_pressure = get_random_float(1.5, 3.5, 1)
    elif gas_type == 'N2':
        gas_pressure = get_random_float(2.5, 5, 1)
    elif gas_type == 'Ar':
        gas_pressure = get_random_float(3.0, 6.0, 1)
    else:
        gas_pressure = get_random_float(1, 2.5, 1)

    # Power and speed
    power = get_random_float(power_min, power_max, 0)
    speed = get_random_float(speed_min, speed_max, 0)

    # Quality distribution: 70% 5-star, 20% 4-star, 10% 3-star
    quality_roll = random.random()
    if quality_roll < 0.7:
        quality = 5
    elif quality_roll < 0.9:
        quality = 4
    else:
        quality = 3

    # Edge quality: 75% clean, 25% slight_dross
    edge_quality = 'clean' if random.random() < 0.75 else 'slight_dross'

    return {
        'material': material,
        'thickness_mm': thickness,
        'power_pct': int(power),
        'speed_mm_min': int(speed),
        'gas_type': gas_type,
        'gas_pressure_bar': gas_pressure,
        'focus_position_mm': focus_position,
        'nozzle_diameter_mm': nozzle_diameter,
        'nozzle_distance_mm': nozzle_distance,
        'line_interval_mm': line_interval,
        'quality_rating': quality,
        'edge_quality': edge_quality,
        'source': 'ai_baseline',
        'is_shared': True,
        'created_at': generate_date_in_last_n_days(30),
    }


def generate_all_parameter_sets() -> List[Dict]:
    """Generate all parameter sets."""
    all_sets = []

    for material, config in MATERIAL_CONFIGS.items():
        for thickness in config['thicknesses']:
            # Generate 5-7 variations per material/thickness combo
            variations = random.randint(5, 7)
            for _ in range(variations):
                all_sets.append(generate_parameter_set(material, thickness, config))

    return all_sets


def generate_sql_inserts(parameter_sets: List[Dict]) -> str:
    """Generate SQL INSERT statements."""
    lines = [
        '-- AI Baseline Laser Cutting Parameters',
        '-- Generated with realistic physics-based defaults',
        '-- Paste into Supabase SQL Editor to insert',
        '',
        'INSERT INTO cuts (material, thickness_mm, power_pct, speed_mm_min, gas_type, gas_pressure_bar, focus_position_mm, nozzle_diameter_mm, nozzle_distance_mm, line_interval_mm, quality_rating, edge_quality, source, is_shared, created_at) VALUES',
    ]

    for idx, set_data in enumerate(parameter_sets):
        created_at = f"'{set_data['created_at']}'::timestamptz"

        values = [
            f"'{set_data['material']}'",
            str(set_data['thickness_mm']),
            str(set_data['power_pct']),
            str(set_data['speed_mm_min']),
            f"'{set_data['gas_type']}'",
            str(set_data['gas_pressure_bar']),
            str(set_data['focus_position_mm']),
            str(set_data['nozzle_diameter_mm']),
            str(set_data['nozzle_distance_mm']),
            str(set_data['line_interval_mm']),
            str(set_data['quality_rating']),
            f"'{set_data['edge_quality']}'",
            "'ai_baseline'",
            'true',
            created_at,
        ]

        comma = ',' if idx < len(parameter_sets) - 1 else ';'
        lines.append(f"({', '.join(values)}){comma}")

    lines.append('')
    return '\n'.join(lines)


def generate_json(parameter_sets: List[Dict]) -> str:
    """Generate JSON output."""
    return json.dumps(parameter_sets, indent=2)


def print_examples(parameter_sets: List[Dict], count: int = 5) -> None:
    """Print sample parameters to console."""
    print(f"\n{'='*80}")
    print(f"SAMPLE BASELINE PARAMETERS (first {count} of {len(parameter_sets)})")
    print(f"{'='*80}\n")

    for i, set_data in enumerate(parameter_sets[:count]):
        print(f"[{i+1}] {set_data['material']} @ {set_data['thickness_mm']}mm")
        print(f"    Power: {set_data['power_pct']}% | Speed: {set_data['speed_mm_min']} mm/min | Gas: {set_data['gas_type']} @ {set_data['gas_pressure_bar']} bar")
        print(f"    Focus: {set_data['focus_position_mm']}mm | Nozzle: {set_data['nozzle_diameter_mm']}mm dia, {set_data['nozzle_distance_mm']}mm dist")
        print(f"    Line interval: {set_data['line_interval_mm']}mm | Quality: {set_data['quality_rating']}★ ({set_data['edge_quality']})")
        print()


def main():
    """Main entry point."""
    args = sys.argv[1:]
    dry_run = '--dry-run' in args
    output_format = 'sql' if '--sql' in args else 'json' if '--json' in args else 'both'

    output_dir = './data'
    if '--output' in args:
        output_dir = args[args.index('--output') + 1]

    print('Generating realistic AI baseline laser cutting parameters...\n')

    # Generate all parameter sets
    parameter_sets = generate_all_parameter_sets()
    print(f'✓ Generated {len(parameter_sets)} parameter sets')

    # Show examples
    print_examples(parameter_sets, 5)

    if dry_run:
        print(f'\n[DRY RUN] Would generate {len(parameter_sets)} records to database.\n')
        return

    # Create output directory
    output_path = Path(output_dir)
    output_path.mkdir(parents=True, exist_ok=True)

    # Generate outputs
    if output_format in ('json', 'both'):
        json_path = output_path / 'baseline-parameters.json'
        json_path.write_text(generate_json(parameter_sets))
        print(f'✓ JSON output: {json_path}')

    if output_format in ('sql', 'both'):
        sql_path = output_path / 'baseline-parameters.sql'
        sql_path.write_text(generate_sql_inserts(parameter_sets))
        print(f'✓ SQL output: {sql_path}')

    print(f'\nUsage:')
    print(f'  python3 scripts/generate-baseline-data.py            # Generate both JSON + SQL')
    print(f'  python3 scripts/generate-baseline-data.py --json     # JSON only')
    print(f'  python3 scripts/generate-baseline-data.py --sql      # SQL only')
    print(f'  python3 scripts/generate-baseline-data.py --dry-run  # Preview only')
    print(f'  python3 scripts/generate-baseline-data.py --output ./custom/path')


if __name__ == '__main__':
    main()

File: scripts/generate-baseline-data.ts
#!/usr/bin/env node

/**
 * Generate realistic AI baseline laser cutting parameters for Supabase seeding
 *
 * Generates ~250-300 parameter sets across common materials and thicknesses
 * with scientifically reasonable defaults based on material properties.
 *
 * This TypeScript version provides type safety and better IDE support.
 */

import * as fs from 'fs';
import * as path from 'path';

// Type definitions
interface PowerRange {
  min: number;
  max: number;
}

interface SpeedRange {
  min: number;
  max: number;
}

interface MaterialConfig {
  category: string;
  thicknesses: number[];
  gasTypes: string[];
  powerRange: (thickness: number) => PowerRange;
  speedRange: (thickness: number) => SpeedRange;
}

interface ParameterSet {
  material: string;
  thickness_mm: number;
  power_pct: number;
  speed_mm_min: number;
  gas_type: string;
  gas_pressure_bar: number;
  focus_position_mm: number;
  nozzle_diameter_mm: number;
  nozzle_distance_mm: number;
  line_interval_mm: number;
  pulse_frequency_hz?: number;
  quality_rating: number;
  edge_quality: string;
  source: string;
  is_shared: boolean;
  created_at: string;
}

interface DatabaseInsertRow extends ParameterSet {
  machine_id: string;
  user_id: string;
}

// Material properties & cutting parameters
// Based on industry standards for fiber lasers
const MATERIAL_CONFIGS: Record<string, MaterialConfig> = {
  'Stainless Steel 304': {
    category: 'stainless',
    thicknesses: [0.5, 1, 2, 3, 5, 8],
    gasTypes: ['O2', 'N2', 'air'],
    powerRange: (thickness: number): PowerRange => ({
      min: Math.max(20, thickness === 0.5 ? 25 : thickness <= 2 ? 40 : 60),
      max: Math.min(100, thickness === 0.5 ? 80 : thickness <= 2 ? 90 : 100),
    }),
    speedRange: (thickness: number): SpeedRange => ({
      min: thickness === 0.5 ? 800 : thickness <= 2 ? 400 : 150,
      max: thickness === 0.5 ? 1400 : thickness <= 2 ? 800 : 400,
    }),
  },
  'Mild Steel': {
    category: 'mild_steel',
    thicknesses: [0.5, 1, 2, 3, 5, 8],
    gasTypes: ['O2', 'air'],
    powerRange: (thickness: number): PowerRange => ({
      min: thickness === 0.5 ? 20 : thickness <= 2 ? 35 : 50,
      max: thickness === 0.5 ? 75 : thickness <= 2 ? 85 : 95,
    }),
    speedRange: (thickness: number): SpeedRange => ({
      min: thickness === 0.5 ? 900 : thickness <= 2 ? 500 : 200,
      max: thickness === 0.5 ? 1600 : thickness <= 2 ? 1000 : 500,
    }),
  },
  'Aluminum': {
    category: 'aluminum',
    thicknesses: [0.5, 1, 2, 3, 5],
    gasTypes: ['N2', 'air'],
    powerRange: (thickness: number): PowerRange => ({
      min: thickness === 0.5 ? 15 : thickness <= 2 ? 30 : 45,
      max: thickness === 0.5 ? 70 : thickness <= 2 ? 80 : 90,
    }),
    speedRange: (thickness: number): SpeedRange => ({
      min: thickness === 0.5 ? 1000 : thickness <= 2 ? 600 : 300,
      max: thickness === 0.5 ? 1800 : thickness <= 2 ? 1200 : 700,
    }),
  },
  'Acrylic': {
    category: 'non_metal',
    thicknesses: [1, 3, 5, 8, 10],
    gasTypes: ['air'],
    powerRange: (thickness: number): PowerRange => ({
      min: thickness <= 3 ? 10 : 15,
      max: thickness <= 3 ? 40 : 60,
    }),
    speedRange: (thickness: number): SpeedRange => ({
      min: thickness <= 3 ? 200 : 100,
      max: thickness <= 3 ? 600 : 300,
    }),
  },
  'Copper': {
    category: 'copper',
    thicknesses: [0.5, 1, 2, 3],
    gasTypes: ['N2'],
    powerRange: (thickness: number): PowerRange => ({
      min: thickness === 0.5 ? 30 : thickness <= 2 ? 50 : 70,
      max: thickness === 0.5 ? 85 : thickness <= 2 ? 95 : 100,
    }),
    speedRange: (thickness: number): SpeedRange => ({
      min: thickness === 0.5 ? 300 : thickness <= 2 ? 150 : 80,
      max: thickness === 0.5 ? 800 : thickness <= 2 ? 400 : 250,
    }),
  },
  'Leather': {
    category: 'non_metal',
    thicknesses: [1, 2, 3, 5],
    gasTypes: ['air'],
    powerRange: (thickness: number): PowerRange => ({
      min: thickness <= 2 ? 5 : 10,
      max: thickness <= 2 ? 25 : 40,
    }),
    speedRange: (thickness: number): SpeedRange => ({
      min: thickness <= 2 ? 100 : 50,
      max: thickness <= 2 ? 400 : 200,
    }),
  },
};

// Helper functions
function getRandomInt(min: number, max: number): number {
  return Math.floor(Math.random() * (max - min + 1)) + min;
}

function getRandomFloat(min: number, max: number, decimals: number = 1): number {
  const num = Math.random() * (max - min) + min;
  return Math.round(num * Math.pow(10, decimals)) / Math.pow(10, decimals);
}

function randomElement<T>(arr: T[]): T {
  return arr[Math.floor(Math.random() * arr.length)];
}

function generateDateInLastNDays(days: number): string {
  const now = new Date();
  const daysAgo = Math.random() * days;
  const date = new Date(now.getTime() - daysAgo * 24 * 60 * 60 * 1000);
  return date.toISOString();
}

function generateParameterSet(material: string, thickness: number, config: MaterialConfig): ParameterSet {
  const powerRange = config.powerRange(thickness);
  const speedRange = config.speedRange(thickness);
  const gasType = randomElement(config.gasTypes);

  // Focus position: typically -5 to +5mm relative to surface
  const focusPosition = getRandomFloat(-5, 5, 1);

  // Line interval: affects scan lines per mm, typically 0.05-0.15mm
  const lineInterval = getRandomFloat(0.05, 0.15, 3);

  // Nozzle diameter: typically 1-4mm
  const nozzleDiameter = getRandomFloat(1, 4, 1);

  // Nozzle distance: typically 0.5-2mm from surface
  const nozzleDistance = getRandomFloat(0.5, 2, 1);

  // Gas pressure varies by gas type
  let gasPressure: number;
  if (gasType === 'O2') {
    gasPressure = getRandomFloat(1.5, 3.5, 1);
  } else if (gasType === 'N2') {
    gasPressure = getRandomFloat(2.5, 5, 1);
  } else {
    gasPressure = getRandomFloat(1, 2.5, 1);
  }

  // Power and speed with realistic variation
  const power = getRandomFloat(powerRange.min, powerRange.max, 0);
  const speed = getRandomFloat(speedRange.min, speedRange.max, 0);

  // Quality distribution: 70% 5-star, 20% 4-star, 10% 3-star (realistic for AI baseline)
  const qualityRoll = Math.random();
  let quality: number;
  if (qualityRoll < 0.7) {
    quality = 5;
  } else if (qualityRoll < 0.9) {
    quality = 4;
  } else {
    quality = 3;
  }

  // Edge quality distribution: 75% clean, 25% slight_dross
  const edgeQuality = Math.random() < 0.75 ? 'clean' : 'slight_dross';

  return {
    material,
    thickness_mm: thickness,
    power_pct: power,
    speed_mm_min: speed,
    gas_type: gasType,
    gas_pressure_bar: gasPressure,
    focus_position_mm: focusPosition,
    nozzle_diameter_mm: nozzleDiameter,
    nozzle_distance_mm: nozzleDistance,
    line_interval_mm: lineInterval,
    quality_rating: quality,
    edge_quality: edgeQuality,
    source: 'ai_baseline',
    is_shared: true,
    created_at: generateDateInLastNDays(30),
  };
}

function generateAllParameterSets(): ParameterSet[] {
  const allSets: ParameterSet[] = [];

  for (const [material, config] of Object.entries(MATERIAL_CONFIGS)) {
    for (const thickness of config.thicknesses) {
      // Generate 2-4 variations per material/thickness combo
      const variations = getRandomInt(2, 4);
      for (let i = 0; i < variations; i++) {
        allSets.push(generateParameterSet(material, thickness, config));
      }
    }
  }

  return allSets;
}

function generateSQLInserts(parameterSets: ParameterSet[]): string {
  const inserts: string[] = [];

  inserts.push(
    '-- AI Baseline Laser Cutting Parameters',
    '-- Generated with realistic physics-based defaults',
    '-- Insert via: psql -h <host> -d <db> -U <user> -f inserts.sql',
    '',
    'INSERT INTO cuts (machine_id, user_id, material, thickness_mm, power_pct, speed_mm_min, gas_type, gas_pressure_bar, focus_position_mm, nozzle_diameter_mm, nozzle_distance_mm, line_interval_mm, quality_rating, edge_quality, is_shared, created_at) VALUES'
  );

  const valueLines = parameterSets.map((set, idx) => {
    // Use a fixed UUID for baseline machine and user for bulk insert
    const machineId = "'550e8400-e29b-41d4-a716-446655440000'::uuid";
    const userId = "'550e8400-e29b-41d4-a716-446655440001'::uuid";
    const createdAt = `'${set.created_at}'::timestamptz`;

    const values = [
      machineId,
      userId,
      `'${set.material}'`,
      set.thickness_mm,
      set.power_pct,
      set.speed_mm_min,
      `'${set.gas_type}'`,
      set.gas_pressure_bar,
      set.focus_position_mm,
      set.nozzle_diameter_mm,
      set.nozzle_distance_mm,
      set.line_interval_mm,
      set.quality_rating,
      `'${set.edge_quality}'`,
      set.is_shared,
      createdAt,
    ];

    const comma = idx < parameterSets.length - 1 ? ',' : ';';
    return `(${values.join(', ')})${comma}`;
  });

  inserts.push(...valueLines);
  inserts.push('');

  return inserts.join('\n');
}

function generateJSON(parameterSets: ParameterSet[]): string {
  return JSON.stringify(parameterSets, null, 2);
}

function printExamples(parameterSets: ParameterSet[], count: number = 5): void {
  console.log(`\n${'='.repeat(80)}`);
  console.log(`SAMPLE BASELINE PARAMETERS (first ${count} of ${parameterSets.length})`);
  console.log(`${'='.repeat(80)}\n`);

  for (let i = 0; i < Math.min(count, parameterSets.length); i++) {
    const set = parameterSets[i];
    console.log(`[${i + 1}] ${set.material} @ ${set.thickness_mm}mm`);
    console.log(`    Power: ${set.power_pct}% | Speed: ${set.speed_mm_min} mm/min | Gas: ${set.gas_type} @ ${set.gas_pressure_bar} bar`);
    console.log(`    Focus: ${set.focus_position_mm}mm | Nozzle: ${set.nozzle_diameter_mm}mm dia, ${set.nozzle_distance_mm}mm dist`);
    console.log(`    Line interval: ${set.line_interval_mm}mm | Quality: ${set.quality_rating}★ (${set.edge_quality})`);
    console.log('');
  }
}

function main(): void {
  const args = process.argv.slice(2);
  const dryRun = args.includes('--dry-run');
  const outputFormat = args.includes('--sql') ? 'sql' : args.includes('--json') ? 'json' : 'both';
  const outputDirIndex = args.indexOf('--output');
  const outputDir = outputDirIndex !== -1 ? args[outputDirIndex + 1] : './data';

  console.log('Generating realistic AI baseline laser cutting parameters...\n');

  // Generate all parameter sets
  const parameterSets = generateAllParameterSets();
  console.log(`✓ Generated ${parameterSets.length} parameter sets`);

  // Show examples
  printExamples(parameterSets, 5);

  if (dryRun) {
    console.log(`\n[DRY RUN] Would generate ${parameterSets.length} records to database.\n`);
    return;
  }

  // Ensure output directory exists
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  // Generate outputs
  if (outputFormat === 'json' || outputFormat === 'both') {
    const jsonPath = path.join(outputDir, 'baseline-parameters.json');
    fs.writeFileSync(jsonPath, generateJSON(parameterSets));
    console.log(`✓ JSON output: ${jsonPath}`);
  }

  if (outputFormat === 'sql' || outputFormat === 'both') {
    const sqlPath = path.join(outputDir, 'baseline-parameters.sql');
    fs.writeFileSync(sqlPath, generateSQLInserts(parameterSets));
    console.log(`✓ SQL output: ${sqlPath}`);
  }

  console.log(`\nUsage:`);
  console.log(`  npx node scripts/generate-baseline-data.ts            # Generate both JSON + SQL`);
  console.log(`  npx node scripts/generate-baseline-data.ts --json     # JSON only`);
  console.log(`  npx node scripts/generate-baseline-data.ts --sql      # SQL only`);
  console.log(`  npx node scripts/generate-baseline-data.ts --dry-run  # Preview only`);
  console.log(`  npx node scripts/generate-baseline-data.ts --output ./custom/path`);
}

main();
